### PR TITLE
Improving MPTrie performance

### PR DESCRIPTION
**Pull request by gennadylaventman**
Improving MPTrie.persistsSubtrie performance.

Signed-off-by: Gennady Laventman <gennady@il.ibm.com>

### CODE REVIEW

File: internal/mptrie/store/open.go [hunk 1]
```diff
@@ -14,6 +14,8 @@ import (
 )
 
 var (
+	nodeGetTotal = 0
+	nodeGetDB = 0
```
Comment (cendhu) [inline internal/mptrie/store/open.go:18]: Please remove. For time, we can calculate and print the time taken. For low level metrics like these, we need to use prometheus.

File: internal/mptrie/store/update_and_query.go [hunk 2]
```diff
@@ -171,6 +171,9 @@ func (s *Store) CommitChanges(blockNum uint64) error {
 	if err := s.trieDataDB.Write(batch, &opt.WriteOptions{Sync: true}); err != nil {
 		return err
 	}
+	s.logger.Infof("mptrie commit - total get ops %d, db get ops %d, put ops %d, persist ops %d", nodeGetTotal, nodeGetDB, len(s.inMemoryNodes), len(s.nodesToPersist))
+	nodeGetTotal = 0
+	nodeGetDB = 0
```
Comment (cendhu) [inline internal/mptrie/store/update_and_query.go:176]: Please remove. You can add time taken to persistTrie but for low level metrics, let's not use logs.

File: internal/mptrie/store/update_and_query.go [hunk 3]
```diff
@@ -118,7 +120,6 @@ func (s *Store) PersistNode(nodePtr []byte) (bool, error) {
 	nodeBytesWithType, ok := s.inMemoryNodes[key]
 	if ok {
 		s.nodesToPersist[key] = nodeBytesWithType
-		delete(s.inMemoryNodes, key)
```
Comment (cendhu) [inline internal/mptrie/store/update_and_query.go:121]: why should we remove this line?

**Review by cendhu — COMMENTED** [review]
I have tested the performance using this PR. The time taken to commit the trie is reduced by 72%. Nice work. 

A few minor comments.

File: internal/mptrie/mptrie.go [hunk 4]
```diff
@@ -454,31 +454,36 @@ func (t *MPTrie) Commit(blockNum uint64) error {
...
 	case *BranchNode:
 		for _, childPtr := range n.Children {
 			if childPtr != nil {
-				child, err := t.store.GetNode(childPtr)
-				if err != nil {
-					return err
-				}
-
-				if err = t.persistSubtrie(child); err != nil {
+				if err = t.persistSubtrie(childPtr); err != nil {
```
Comment (C0rWin) [inline internal/mptrie/mptrie.go:486]: While you are already at it, can you please reverse the logic of if statement above to reduce code indentation?

File: internal/mptrie/store/open.go (hunk 1 above)
Comment (gennadylaventman) [reply internal/mptrie/store/open.go:18]: ok

File: internal/mptrie/store/update_and_query.go (hunk 2 above)
Comment (gennadylaventman) [reply internal/mptrie/store/update_and_query.go:176]: ok

File: internal/mptrie/store/update_and_query.go (hunk 3 above)
Comment (gennadylaventman) [reply internal/mptrie/store/update_and_query.go:121]: It doesn't affect anything - we will clear this map in `CommitChanges` call.

File: internal/mptrie/mptrie.go (hunk 4 above)
Comment (gennadylaventman) [reply internal/mptrie/mptrie.go:486]: Done, but it adds extra code line :)

Comment (gennadylaventman) [reply internal/mptrie/mptrie.go]: Done

**Review by cendhu — APPROVED** [review]

**Pull request merged**